controllers/dnspolicy_status_updater.go

```diff
@@ -0,0 +1,121 @@
+}
+
+func (r *DNSPolicyStatusUpdater) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.updateStatus,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.DNSPolicyGroupKind},
+			{Kind: &DNSRecordGroupKind},
+		},
+	}
+}
+
+func (r *DNSPolicyStatusUpdater) updateStatus(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("DNSPolicyStatusUpdater")
+
+	policyTypeFilterFunc := dnsPolicyTypeFilterFunc()
+	policyAcceptedFunc := dnsPolicyAcceptedStatusFunc(state)
+	policyErrorFunc := dnsPolicyErrorFunc(state)
+
+	policies := lo.FilterMap(topology.Policies().Items(), policyTypeFilterFunc)
+
+	logger.V(1).Info("updating dns policy statuses", "policies", len(policies))
+
+	for _, policy := range policies {
+		pLogger := logger.WithValues("policy", policy.GetLocator())
+
+		pLogger.V(1).Info("updating dns policy status")
+
+		if policy.GetDeletionTimestamp() != nil {
+			pLogger.V(1).Info("policy marked for deletion, skipping")
+			continue
+		}
+
+		// copy initial conditions, otherwise status will always be updated
+		newStatus := &kuadrantv1alpha1.DNSPolicyStatus{
+			Conditions:         slices.Clone(policy.Status.Conditions),
+			ObservedGeneration: policy.Status.ObservedGeneration,
+		}
+
+		accepted, err := policyAcceptedFunc(policy)
+		meta.SetStatusCondition(&newStatus.Conditions, *kuadrant.AcceptedCondition(policy, err))
+
+		// do not set enforced condition if Accepted condition is false
+		if !accepted {
+			meta.RemoveStatusCondition(&newStatus.Conditions, string(kuadrant.PolicyConditionEnforced))
+		} else {
+			policyRecords := lo.FilterMap(topology.Objects().Children(policy), func(item machinery.Object, _ int) (*kuadrantdnsv1alpha1.DNSRecord, bool) {
+				if rObj, isObj := item.(*controller.RuntimeObject); isObj {
+					if record, isRec := rObj.Object.(*kuadrantdnsv1alpha1.DNSRecord); isRec {
+						return record, true
+					}
+				}
+				return nil, false
+			})
+
+			enforcedCond := enforcedCondition(policyRecords, policy)
+			if pErr := policyErrorFunc(policy); pErr != nil {
+				pLogger.V(1).Info("adding contextual error to policy enforced status", "err", pErr)
+				enforcedCond.Message = fmt.Sprintf("%s : %s", enforcedCond.Message, pErr.Error())
+			}
+			meta.SetStatusCondition(&newStatus.Conditions, *enforcedCond)
+
+			propagateRecordConditions(policyRecords, newStatus)
+
+			newStatus.TotalRecords = int32(len(policyRecords))
+		}
+
+		equalStatus := equality.Semantic.DeepEqual(newStatus, policy.Status)
+		if equalStatus && policy.Generation == policy.Status.ObservedGeneration {
+			pLogger.V(1).Info("policy status unchanged, skipping update")
+			continue
+		}
+		newStatus.ObservedGeneration = policy.Generation
+		policy.Status = *newStatus
+
+		obj, err := controller.Destruct(policy)
```

Comment:
does this remove it from the topology and then it gets re-added when we update it?

Reply:
If you are referring to `controller.Destruct(policy)`, this is just converting it to a `unstructured.Unstructured` type so it works with the client https://github.com/Kuadrant/policy-machinery/blob/main/controller/runnable.go#L231-L237

I found this method name confusing when i saw it being used also.